src/mapping/nav2_map_server/src/map_reps/occupancy_grid.cpp

```diff
@@ -0,0 +1,324 @@
+  const char * name = map_name_.c_str();
+  unsigned char * pixels;
+  unsigned char * p;
+  unsigned char value;
+  int rowstride, n_channels, avg_channels;
+  unsigned int i, j;
+  int k;
+  double occ;
+  int alpha;
+  int color_sum;
+  double color_avg;
+
+  // Load the image using SDL.  If we get NULL back, the image load failed.
+  if (!(img = IMG_Load(name))) {
+    std::string errmsg = std::string("failed to open image file \"") +
+      std::string(name) + std::string("\": ") + IMG_GetError();
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"), "%s", errmsg.c_str());
+
+    throw std::runtime_error(errmsg);
+  }
+  // Copy the image data into the map structure
+  map_msg_.info.width = img->w;
+  map_msg_.info.height = img->h;
+  map_msg_.info.resolution = res;
+  map_msg_.info.origin.position.x = *(origin);
+  map_msg_.info.origin.position.y = *(origin + 1);
+  map_msg_.info.origin.position.z = 0.0;
+  btQuaternion q;
+  // setEulerZYX(yaw, pitch, roll)
+  q.setEulerZYX(*(origin + 2), 0, 0);
+  map_msg_.info.origin.orientation.x = q.x();
+  map_msg_.info.origin.orientation.y = q.y();
+  map_msg_.info.origin.orientation.z = q.z();
+  map_msg_.info.origin.orientation.w = q.w();
+
+  // Allocate space to hold the data
+  map_msg_.data.resize(map_msg_.info.width * map_msg_.info.height);
+
+  // Get values that we'll need to iterate through the pixels
+  rowstride = img->pitch;
+  n_channels = img->format->BytesPerPixel;
+
+  // NOTE: Trinary mode still overrides here to preserve existing behavior.
+  // Alpha will be averaged in with color channels when using trinary mode.
+  if (mode == TRINARY || !img->format->Amask) {
+    avg_channels = n_channels;
+  } else {
+    avg_channels = n_channels - 1;
+  }
+
+  // Copy pixel data into the map structure
+  pixels = (unsigned char *)(img->pixels);
+  for (j = 0; j < map_msg_.info.height; j++) {
+    for (i = 0; i < map_msg_.info.width; i++) {
+      // Compute mean of RGB for this pixel
+      p = pixels + j * rowstride + i * n_channels;
+      color_sum = 0;
+      for (k = 0; k < avg_channels; k++) {
+        color_sum += *(p + (k));
+      }
+      color_avg = color_sum / static_cast<double>(avg_channels);
+
+      if (n_channels == 1) {
+        alpha = 1;
+      } else {
+        alpha = *(p + n_channels - 1);
+      }
+      if (negate) {
+        color_avg = 255 - color_avg;
+      }
+      if (mode == RAW) {
+        value = color_avg;
+        map_msg_.data[MAP_IDX(map_msg_.info.width, i, map_msg_.info.height - j - 1)] = value;
+        continue;
+      }
+
+      // If negate is true, we consider blacker pixels free, and whiter
+      // pixels free.  Otherwise, it's vice versa.
+      occ = (255 - color_avg) / 255.0;
+
+      // Apply thresholds to RGB means to determine occupancy values for
+      // map.  Note that we invert the graphics-ordering of the pixels to
+      // produce a map with cell (0,0) in the lower-left corner.
+      if (occ > occ_th) {
+        value = +100;
```

Comment:
I'd recommend making the +100, 0, and -1 values class constant variables.

Reply:
Will consider redesign in future pull request

Reply:
Pretty small change IMO so I'd reconsider doing it now.  Otherwise please make a ticket or at least a TODO.